test/prototype_common_utils.py

```diff
@@ -34,26 +133,104 @@ def __iter__(self):
         yield self.args
         yield self.kwargs
 
-    def __str__(self):
-        def short_repr(obj, max=20):
-            repr_ = repr(obj)
-            if len(repr_) <= max:
-                return repr_
+    def load(self, device="cpu"):
+        args = tuple(arg.load(device) if isinstance(arg, TensorLoader) else arg for arg in self.args)
+        kwargs = {
+            keyword: arg.load(device) if isinstance(arg, TensorLoader) else arg for keyword, arg in self.kwargs.items()
+        }
+        return args, kwargs
 
-            return f"{repr_[:max//2]}...{repr_[-(max//2-3):]}"
+    def __repr__(self):
+        def better_repr(obj):
+            if isinstance(obj, enum.Enum):
+                return str(obj)
+            else:
+                return repr(obj)
 
         return ", ".join(
             itertools.chain(
-                [short_repr(arg) for arg in self.args],
-                [f"{param}={short_repr(kwarg)}" for param, kwarg in self.kwargs.items()],
+                [better_repr(arg) for arg in self.args],
+                [f"{param}={better_repr(kwarg)}" for param, kwarg in self.kwargs.items()],
             )
         )
 
 
-make_tensor = functools.partial(torch.testing.make_tensor, device="cpu")
+DEFAULT_SQUARE_IMAGE_SIZE = (16, 16)
+DEFAULT_LANDSCAPE_IMAGE_SIZE = (7, 33)
+DEFAULT_PORTRAIT_IMAGE_SIZE = (31, 9)
+DEFAULT_IMAGE_SIZES = (DEFAULT_LANDSCAPE_IMAGE_SIZE, DEFAULT_PORTRAIT_IMAGE_SIZE, DEFAULT_SQUARE_IMAGE_SIZE)
+
+DEFAULT_EXTRA_DIMS = ((), (0,), (4,), (2, 3), (5, 0), (0, 5))
+
+
+def from_loader(loader_fn):
+    def wrapper(*args, **kwargs):
+        loader = loader_fn(*args, **kwargs)
+        return loader.load(kwargs.get("device", "cpu"))
+
+    return wrapper
+
+
+def from_loaders(loaders_fn):
+    def wrapper(*args, **kwargs):
+        loaders = loaders_fn(*args, **kwargs)
+        for loader in loaders:
+            yield loader.load(kwargs.get("device", "cpu"))
+
+    return wrapper
+
+
+class TensorLoader:
+    def __init__(self, fn, *, shape, dtype):
+        self.fn = fn
+        self.shape = shape
+        self.dtype = dtype
+
+    def unwrap(self):
+        return TensorLoader(
+            lambda shape, dtype, device: torch.Tensor(self.fn(shape, dtype, device)), shape=self.shape, dtype=self.dtype
+        )
+
+    def load(self, device):
+        return self.fn(self.shape, self.dtype, device)
+
+    _TYPE_NAME = "torch.Tensor"
 
+    def _extra_repr(self):
+        return []
 
-def make_image(size=None, *, color_space, extra_dims=(), dtype=torch.float32, constant_alpha=True):
+    def __repr__(self):
+        extra = ", ".join(
+            [
+                str(tuple(self.shape)),
+                str(self.dtype).replace("torch.", ""),
+                *[str(extra) for extra in self._extra_repr()],
+            ]
+        )
+        return f"{self._TYPE_NAME}[{extra}]"
```

Comment:
This is purely to have more readable test names. I would prefer to keep it, but removing this doesn't affect the framework proposed in this PR.

Reply:
I've refactored this to avoid overly complicated test ids. They distracted more than being helpful.